Remove LTO flags to prevent errors with clang?

Error in question: 

```cpp
pythran -v mymod.py -o mymod_default.so
/home/avmo/src/fluidmeta/pythran/pythran/interval.py:90: RuntimeWarning: invalid value encountered in double_scalars
  return Interval(self.low + other.low, self.high + other.high)
running build_ext
new_compiler returns <class 'distutils.unixccompiler.UnixCCompiler'>
building 'mymod_default' extension
C compiler: clang++ -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fstack-protector-strong -fno-plt -flto=4 -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -march=x86-64 -mtune=generic -O3 -pipe -fstack-protector-strong -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fstack-protector-strong -fno-plt -fPIC
creating /tmp/tmptddrj69d/tmp
compile options: '-DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -DPYTHRAN_BLAS_BLAS -I/home/avmo/src/fluidmeta/pythran/pythran -I/home/avmo/src/fluidmeta/pythran/pythran -I/home/avmo/src/fluidmeta/pythran/pythran/pythonic/patch -I/scratch/avmo/opt/fluidmeta/lib/python3.7/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I/scratch/avmo/opt/fluidmeta/include -I/home/avmo/src/fluidmeta/pythran/pythran/pythonic/patch -I/scratch/avmo/opt/fluidmeta/lib/python3.7/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I/scratch/avmo/opt/fluidmeta/include -I/usr/include/python3.7m -c'
extra options: '-std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv'
clang++: /tmp/tmpfwy7l6cs.cpp
exec_command(['clang++', '-pthread', '-Wno-unused-result', '-Wsign-compare', '-DNDEBUG', '-g', '-fwrapv', '-O3', '-Wall', '-march=x86-64', '-mtune=generic', '-O3', '-pipe', '-fstack-protector-strong', '-fno-plt', '-flto=4', '-fuse-linker-plugin', '-ffat-lto-objects', '-flto-partition=none', '-march=x86-64', '-mtune=generic', '-O3', '-pipe', '-fstack-protector-strong', '-fno-plt', '-march=x86-64', '-mtune=generic', '-O3', '-pipe', '-fstack-protector-strong', '-fno-plt', '-fPIC', '-DENABLE_PYTHON_MODULE', '-D__PYTHRAN__=3', '-DENABLE_PYTHON_MODULE', '-D__PYTHRAN__=3', '-DPYTHRAN_BLAS_BLAS', '-DPYTHRAN_BLAS_BLAS', '-I/home/avmo/src/fluidmeta/pythran/pythran', '-I/home/avmo/src/fluidmeta/pythran/pythran', '-I/home/avmo/src/fluidmeta/pythran/pythran/pythonic/patch', '-I/scratch/avmo/opt/fluidmeta/lib/python3.7/site-packages/numpy/core/include', '-I/usr/local/include', '-I/usr/include', '-I/scratch/avmo/opt/fluidmeta/include', '-I/home/avmo/src/fluidmeta/pythran/pythran/pythonic/patch', '-I/scratch/avmo/opt/fluidmeta/lib/python3.7/site-packages/numpy/core/include', '-I/usr/local/include', '-I/usr/include', '-I/scratch/avmo/opt/fluidmeta/include', '-I/usr/include/python3.7m', '-c', '/tmp/tmpfwy7l6cs.cpp', '-o', '/tmp/tmptddrj69d/tmp/tmpfwy7l6cs.o', '-std=c++11', '-fno-math-errno', '-w', '-fvisibility=hidden', '-fno-wrapv', '-std=c++11', '-fno-math-errno', '-w', '-fvisibility=hidden', '-fno-wrapv'],)
Retaining cwd: /home/avmo/src/fluidmeta/fluidfft/bench/pythran_implicit_loop
_preserve_environment([])
_update_environment(...)
clang-6.0: error: unknown argument: '-flto-partition=none'
clang-6.0: error: unsupported argument '4' to option 'flto='
_update_environment(...)
clang-6.0: error: unknown argument: '-flto-partition=none'
clang-6.0: error: unsupported argument '4' to option 'flto='
WARNING  Compilation error, trying hard to find its origin...
WARNING  Nop, I'm going to flood you with C++ errors!
CRITICAL Cover me Jack. Jack? Jaaaaack!!!!
E: error: Command "clang++ -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fstack-protector-strong -fno-plt -flto=4 -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -march=x86-64 -mtune=generic -O3 -pipe -fstack-protector-strong -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fstack-protector-strong -fno-plt -fPIC -DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DENABLE_PYTHON_MODULE -D__PYTHRAN__=3 -DPYTHRAN_BLAS_BLAS -DPYTHRAN_BLAS_BLAS -I/home/avmo/src/fluidmeta/pythran/pythran -I/home/avmo/src/fluidmeta/pythran/pythran -I/home/avmo/src/fluidmeta/pythran/pythran/pythonic/patch -I/scratch/avmo/opt/fluidmeta/lib/python3.7/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I/scratch/avmo/opt/fluidmeta/include -I/home/avmo/src/fluidmeta/pythran/pythran/pythonic/patch -I/scratch/avmo/opt/fluidmeta/lib/python3.7/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I/scratch/avmo/opt/fluidmeta/include -I/usr/include/python3.7m -c /tmp/tmpfwy7l6cs.cpp -o /tmp/tmptddrj69d/tmp/tmpfwy7l6cs.o -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv -std=c++11 -fno-math-errno -w -fvisibility=hidden -fno-wrapv" failed with exit status 1
make: *** [Makefile:10: mymod_default.so] Error 1
```

I had to patch as follows to allow pythran to use clang.

```patch
diff --git a/pythran/dist.py b/pythran/dist.py
index 5ad12e9c..b949734f 100644
--- a/pythran/dist.py
+++ b/pythran/dist.py
@@ -71,10 +71,12 @@ class PythranBuildExt(LegacyBuildExt, object):
         # In general, distutils uses -Wstrict-prototypes, but this option
         # is not valid for C++ code, only for C.  Remove it if it's there
         # to avoid a spurious warning on every compilation.
-        try:
-            self.compiler.compiler_so.remove('-Wstrict-prototypes')
-        except (AttributeError, ValueError):
-            pass
+        for flag in ('-Wstrict-prototypes', '-flto-partition=none', '-flto=4'):
+            try:
+                self.compiler.compiler_so.remove(flag)
+                self.compiler.linker_so.remove(flag)
+            except (AttributeError, ValueError):
+                pass
         # Remove -arch i386 if 'x86_64' is specified, otherwise incorrect
         # code is generated, at least on OSX
```

The reason was my `python-config` supplied those flags
```sh
~/s/f/pythran ❯❯❯ python-config --ldflags
 -L/usr/lib -lpython3.7m -lpthread -ldl  -lutil -lm  -Xlinker -export-dynamic
~/s/f/pythran ❯❯❯ python-config --cflags
-I/usr/include/python3.7m -I/usr/include/python3.7m  -Wno-unused-result -Wsign-compare -march=x86-64 -mtune=generic -O3 -pipe -fstack-protector-strong -fno-plt -flto=4 -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -DNDEBUG -g -fwrapv -O3 -Wall
```

I understand this is system specific (Arch Linux), and it does not happen for eg. with python provided by Anaconda.

Also see: https://stackoverflow.com/questions/32237647/how-can-i-link-files-compiled-with-clang-flto-using-lld

--- a/docs/MANUAL.rst
+++ b/docs/MANUAL.rst
@@ -455,11 +455,11 @@ This section contains compiler flags configuration. For education purpose, the d
 
 :``include_dirs``:
 
-    Additionnal include directories to search for headers.
+    Additional include directories to search for headers.
 
 :``cflags``:
 
-    Additionnal random compiler flags (``-f``, ``-O``). Optimization flags generally
+    Additional random compiler flags (``-f``, ``-O``). Optimization flags generally
     go there. The default is to set ``-std=c++11`` for C++11 support.
 
 :``libs``:
@@ -474,11 +474,18 @@ This section contains compiler flags configuration. For education purpose, the d
 
 :``ldflags``:
 
-    Additionnal random linker flags.
+    Additional random linker flags.
 
 :``blas``:
 
     BLAS library to use. Default is ``blas``, but ``atlas`` or ``mkl`` are also viable choices.
+
+:``ignoreflags``:
+
+    Space-separated list of compiler flags that should not be forwarded to the
+    pythran backend compiler when inherited, for instance, from
+    ``python-config``. For instance ``-Wstrict-prototypes`` is a C-only option
+    that should be pruned.
 
 
 ``[pythran]``

--- a/pythran/dist.py
+++ b/pythran/dist.py
@@ -24,18 +24,17 @@ class PythranBuildExt(LegacyBuildExt, object):
     """
     def build_extension(self, ext):
         StringTypes = (str, unicode) if sys.version_info[0] == 2 else (str,)
+
         def get_value(obj, key):
             var = getattr(obj, key)
-            if isinstance(var, Iterable) and \
-                not isinstance(var, StringTypes):
+            if isinstance(var, Iterable) and not isinstance(var, StringTypes):
                 return var[0]
             else:
                 return var
 
         def set_value(obj, key, value):
             var = getattr(obj, key)
-            if isinstance(var, Iterable) and \
-                not isinstance(var, StringTypes):
+            if isinstance(var, Iterable) and not isinstance(var, StringTypes):
                 var[0] = value
             else:
                 setattr(obj, key, value)
@@ -71,10 +70,12 @@ class PythranBuildExt(LegacyBuildExt, object):
         # In general, distutils uses -Wstrict-prototypes, but this option
         # is not valid for C++ code, only for C.  Remove it if it's there
         # to avoid a spurious warning on every compilation.
-        try:
-            self.compiler.compiler_so.remove('-Wstrict-prototypes')
-        except (AttributeError, ValueError):
-            pass
+        for flag in cfg.cfg.get('compiler', "ignoreflags").split():
+            for target in ('compiler_so', 'linker_so'):
+                try:
+                    getattr(self.compiler, target).remove(flag)
+                except (AttributeError, ValueError):
+                    pass
 
         # Remove -arch i386 if 'x86_64' is specified, otherwise incorrect
         # code is generated, at least on OSX

--- a/pythran/pythran-darwin.cfg
+++ b/pythran/pythran-darwin.cfg
@@ -9,3 +9,4 @@ ldflags=
 blas=blas
 CC=
 CXX=
+ignoreflags=-Wstrict-prototypes

--- a/pythran/pythran-linux.cfg
+++ b/pythran/pythran-linux.cfg
@@ -9,3 +9,4 @@ ldflags=-fvisibility=hidden -Wl,-strip-all
 blas=blas
 CC=
 CXX=
+ignoreflags=-Wstrict-prototypes

--- a/pythran/pythran-linux2.cfg
+++ b/pythran/pythran-linux2.cfg
@@ -9,3 +9,4 @@ ldflags=-fvisibility=hidden -Wl,-strip-all
 blas=blas
 CC=
 CXX=
+ignoreflags=-Wstrict-prototypes

--- a/pythran/pythran-win32.cfg
+++ b/pythran/pythran-win32.cfg
@@ -9,3 +9,4 @@ ldflags=
 blas=blas
 CC=
 CXX=
+ignoreflags=

